_posts/en/newsletters/2019-05-14-newsletter.md

```diff
@@ -0,0 +1,485 @@
+[BIP32][] HD key derivation) in order to derive the *taproot output
+key---*the key that is used in any bech32 addresses and scriptPubKeys
+that pay these three conditions.
+
+    FIXME: diagram of tweaking, similar to one from BitMex article
+
+When it comes time to spend this money, either Alice or Bob can provide
+the script they want to use, the data needed by it (a signature and,
+sometimes, a hash preimage), the Taproot internal key, and the hash of
+the script they didn't use.  Alternatively, Alice and Bob work together
+to use signature aggregation (after accounting for the tweak) to provide
+a signature in combination with the same single-key, single-signature
+spending pattern described in the previous subsections.  As long as the
+data they provide in either case is correct, the spend will be accepted.
+
+    FIXME: table example of HTLC cost in current P2WSH and taproot for
+    each of the three possible outcomes
+
+    P2WSH
+      Output: 8+1+34
+      1.preimage, digest, A pubkey, B pubkey, OP_IF/ENDIF, Alice sig
+      2. digest, A pubkey, B pubkey, IF/ENDIF, Bob sig
+      3. N/A (use one of the above)
+
+    Taproot
+      Output: 8+1+35
+      1. preimage, digest, A pubkey, A sig; internal key; leaf hash
+      2. B pubkey, B sig; internal key; leaf hash
+      3. AB sig
+
+Because we chose to use a small tree in the above example, we will
+describe some things that would've been clearer with a larger example.
+First, if you have any unpaired nodes in your tree, your software will
+be responsible for pairing them with something, e.g a 32-byte random
+number.  Additionally, the maximum depth of a tree is 32 rows, allowing
+a tree to contain a maximum of around four billion scripts.  However,
+only one of those scripts would appear in any spending transaction and
+only 32 hashes directly related to the merkle tree would need to be
+generated to prove the script was part of the tree---this means that more
+complex scripts than our simple HTLC can gain much larger efficiency
+improvements than seen above (see an article focused on [MAST][] for
+more information).
+
+### Slight changes to scripting with Tapscript
+
+When using Bitcoin Script with Taproot, some of the rules have been
+changed, most notably:
+
+- **Unused and invalid opcodes redefined:** the bytes of opcodes that
+  were never assigned, or which were made invalid by Satoshi Nakamoto,
+  or which were created for upgrades and have not been used yet have all
+  been changed to have the semantics of an `OP_SUCCESS` opcode that
+  unconditionally renders any script containing that code valid.
+  Because future soft forks can only add new rules by making
+  previously-valid things invalid, this maximal validity
+  today will allow maximal flexibility in future soft forks.  Of course,
+  receivers get to choose what scripts they receive payments to, so no
+  one who likes their money will include an `OP_SUCCESSn` opcode in
```

Comment:
It's OP_SUCCESSx in bip-tapscript as a placeholder for OP_SUCCESS80, OP_SUCCESS98, OP_SUCCESS126, etc.

Reply:
Although I think it was technically correct, I ended up removing the `n` from this text because it was a bit confusing in the reduced context of a summary.